engineering-decision-records/010-ledger-validation-context.md

```diff
@@ -0,0 +1,321 @@
+            // ...
+            // some validations
+            // ...
+
+            context.retire(id, epoch);
+
+            Ok(())
+        }
+
+        _ => ...
+    }
+}
+```
+
+### Lookup context & preparation
+
+In addition to the state-management interface above, we sometimes require read-only lookups of specific entities. So, for all traits above, we define an extra `lookup` method and call the overall a _Slice_. For example, the `UtxoSlice` resembles:
+
+```rust
+pub trait UtxoSlice {
+    fn lookup(&self, input: &TransactionInput) -> Option<&TransactionOutput>;
+    fn consume(&mut self, input: TransactionInput);
+    fn produce(&mut self, input: TransactionInput, output: TransactionOutput);
+}
+```
+
+To satisfy the design constraint (3) above, we assume the slices have prefetched data and aren't fetching data from an on-disk store on the fly. This presents a challenge: how can we ensure the data is available when the `lookup` occurs?
+
+To overcome this, we introduce an extra step which we call _preparation_, and whose role is to traverse the block once to identify which entities will be needed by the validation rules, thus ensuring that the data needed down the line is identified and fetched from the storage in a single batch. Note that the validation context is bounded because blocks are bounded in size. So, the amount of data that needs to be prefetched is limited and acceptable to live transiently in memory during validations. We thus define the following additional traits:
+
+```rust
+pub trait PrepareUtxoSlice<'a> {
+    fn require_input(&'_ mut self, input: &'a TransactionInput);
+}
+
+pub trait PreparePoolsSlice<'a> {
+    fn require_pool(&'_ mut self, pool: &'a PoolId);
+}
+
+pub trait PrepareAccountsSlice<'a> {
+    fn require_account(&'_ mut self, credential: &'a StakeCredential);
+}
+
+pub trait PrepareDRepsSlice<'a> {
+    fn require_drep(&'_ mut self, credential: &'a StakeCredential);
+}
+```
+
+With this, we can perform a single preparation step, which then leads to the creation of a validation slice that reflects a subset (a.k.a a slice) of the ledger state relevant to the transaction. For example, one can identify necessary inputs to lookup as such:
+
+```rust
+pub fn prepare_block<'block>(
+    context: &mut impl PreparationContext<'block>,
+    block: &'block MintedBlock<'_>,
+) {
+    block.transaction_bodies.iter().for_each(|transaction| {
+        let inputs = transaction.inputs.iter();
+
+        let collaterals = transaction
+            .collateral
+            .as_deref()
+            .map(|xs| xs.as_slice())
+            .unwrap_or(&[])
+            .iter();
+
+        let reference_inputs = transaction
+            .reference_inputs
+            .as_deref()
+            .map(|xs| xs.as_slice())
+            .unwrap_or(&[])
+            .iter();
+
+        inputs
+            .chain(reference_inputs)
+            .chain(collaterals)
+            .for_each(|input| context.require_input(input));
```

Comment:
We could consider having some data representation of those requirements instead of fn calls? 
So this function would return a list of e.g. `RerquiredTransactionInput` instead of calling `context.require_input`

Reply:
I was under the impression that this type of more functional pattern were quite idiomatic in Rust? Hence the method call. 

Fundamentally, they are equivalent to me; what would we gain from returning a structured type?

Reply:
I guess it depends what those functions do :) 
Data convey semantic while function are related to computation. Seems easier to me to use data to describe intent, then have implementation be free to interpret those (e.g. here de-duplication, reordering, caching). Strictly speaking you can achieve something similar with traits I guess.

It might be my clojure tropism ;) [This](https://ericnormand.me/article/data-functions-macros-why) is worth reading by the way.

Reply:
Isn't this doing the same thing? The context owns the required inputs, so `require_input` for each input *is* constructing said list, if you want to think about it like that.

I don't come from a functional background, and I'm also not an incredibly experienced rust-acean, but this feels much more natural to me, since the context has to evolve during validation.